rust-runtime/aws-smithy-types/src/timeout.rs

```diff
@@ -0,0 +1,261 @@
+    /// If you want to set a timeout for an entire request including retry attempts,
+    /// use [`TimeoutConfig::api_call_timeout`] instead.
+    pub fn api_call_attempt_timeout(&self) -> Option<Duration> {
+        self.api_call_attempt_timeout
+    }
+
+    /// A limit on the amount of time it takes for request to complete. A single request may be
+    /// comprised of several attemps depending on an app's [`RetryConfig`](super::retry::RetryConfig). If you want
+    /// to control timeouts for a single attempt, use [`TimeoutConfig::api_call_attempt_timeout`].
+    pub fn api_call_timeout(&self) -> Option<Duration> {
+        self.api_call_timeout
+    }
+
+    /// Consume a `TimeoutConfig` to create a new one, setting the connect timeout
+    pub fn with_connect_timeout(mut self, timeout: Option<Duration>) -> Self {
+        self.connect_timeout = timeout;
+        self
+    }
+
+    /// Consume a `TimeoutConfig` to create a new one, setting the TLS negotiation timeout
+    pub fn with_tls_negotiation_timeout(mut self, timeout: Option<Duration>) -> Self {
+        self.tls_negotiation_timeout = timeout;
+        self
+    }
+
+    /// Consume a `TimeoutConfig` to create a new one, setting the read timeout
+    pub fn with_read_timeout(mut self, timeout: Option<Duration>) -> Self {
+        self.read_timeout = timeout;
+        self
+    }
+
+    /// Consume a `TimeoutConfig` to create a new one, setting the API call attempt timeout
+    pub fn with_api_call_attempt_timeout(mut self, timeout: Option<Duration>) -> Self {
+        self.api_call_attempt_timeout = timeout;
+        self
+    }
+
+    /// Consume a `TimeoutConfig` to create a new one, setting the API call timeout
+    pub fn with_api_call_timeout(mut self, timeout: Option<Duration>) -> Self {
+        self.api_call_timeout = timeout;
+        self
+    }
+
+    /// Merge two builders together. Values from `other` will only be used as a fallback for values
+    /// from `self`. Useful for merging configs from different sources together when you want to
+    /// handle "precedence" per value instead of at the config level
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # use std::time::Duration;
+    /// # use aws_smithy_types::timeout::TimeoutConfig;
+    /// let a = TimeoutConfig::new().with_read_timeout(Some(Duration::from_secs(2)));
+    /// let b = TimeoutConfig::new()
+    ///     .with_read_timeout(Some(Duration::from_secs(10)))
+    ///     .with_connect_timeout(Some(Duration::from_secs(3)));
+    /// let timeout_config = a.merge_with(b);
+    /// // A's value take precedence over B's value
+    /// assert_eq!(timeout_config.read_timeout(), Some(Duration::from_secs(2)));
+    /// // A never set a connect timeout so B's value was used
+    /// assert_eq!(timeout_config.connect_timeout(), Some(Duration::from_secs(3)));
+    /// ```
+    pub fn merge_with(self, other: Self) -> Self {
```

Comment:
Might want to call this `take_unset_from` rather than `merge_with` since it's taking the values it doesn't have yet from the other.